src/django_otp/tests.py

```diff
@@ -164,6 +164,78 @@ def test_verify_is_allowed(self):
             self.assertEqual(data3, None)
 
 
+class CooldownTestMixin:
+    def setUp(self):
+        self.device = None
+
+    def valid_token(self):
+        """Returns a valid token to pass to our device under test."""
+        raise NotImplementedError()
+
+    def invalid_token(self):
+        """Returns an invalid token to pass to our device under test."""
+        raise NotImplementedError()
+
+    #
+    # Tests
+    #
+
+    def test_generate_is_allowed_on_first_try(self):
+        """Token generation should be allowed on first try."""
+        allowed, _ = self.device.generate_is_allowed()
+        self.assertTrue(allowed)
+
+    def test_cooldown_imposed_after_successful_generation(self):
+        """
+        Token generation before cooldown should not be allowed
+        and the relevant reason should be returned.
+        """
+        with freeze_time():
+            self.device.generate_challenge()
+            allowed, details = self.device.generate_is_allowed()
```

Comment:
If you insert `self.device.refresh_from_db()` between these lines, the test fails (`last_generated_timestamp` never gets committed).